## 1. Importing libraries

In [45]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 2. Importing data

In [46]:
# Turn project folder path into a string
path = r'C:\Users\Admin\Desktop\Data Analysis\Python\04-2024 Instacart Basket Analysis'

In [47]:
path

'C:\\Users\\Admin\\Desktop\\Data Analysis\\Python\\04-2024 Instacart Basket Analysis'

In [48]:
# Import orders.csv using list
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [49]:
# Import product.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [50]:
# Importing data set departments.csv
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [51]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## 3. Data Wrangling

In [52]:
# Dropping eval_set column from orders.csv
df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [53]:
df_ords_2.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [54]:
# Print all functions within the 'days_since_prior_order' column
df_ords_2['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [55]:
# Rename 'order_dow' columns
df_ords_2.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [56]:
df_ords_2.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [57]:
# Changing data type of the column "order_id" to string from orders.csv
df_ords_2['order_id'] = df_ords_2['order_id'].astype('str')

In [58]:
# Checking data type of the 'order_id' column
df_ords_2['order_id'].dtype

dtype('O')

In [59]:
# Transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [60]:
# Creating a new transposed df_dep dataframe
df_dep_t = df_dep.T

In [61]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [62]:
# Creating an index for df_dep_t dataframe
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [63]:
## Creating a header with the 1st line of df_dep_t dataframe
new_header = df_dep_t.iloc[0]

In [64]:
new_header

0    department
Name: department_id, dtype: object

In [65]:
# Take the data under the header row for a new df
df_dep_t_new = df_dep_t[1:]

In [66]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [67]:
#Set the header row as the df header
df_dep_t_new.columns = new_header

In [68]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## 4. Creating a data dictionary

In [69]:
# Creating a data dictionary from df_dep_t_new that tells what number each department represent
data_dict = df_dep_t_new.to_dict('index')

In [70]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [71]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [72]:
# Checking the data dictionary for what department 19 represent
print(data_dict.get('19'))

{'department': 'snacks'}


## 5. Subsetting

In [73]:
# Create a df_snakcs dataframe with only product with 'department_id' of 19
df_snacks = df_prods[df_prods['department_id'] == 19]

In [74]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

## 6. Task

### 2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [76]:
## Changing 'user_id' column to string
df_ords_2['user_id'] = df_ords_2['user_id'].astype('str')

In [77]:
df_ords_2['user_id'].dtype

dtype('O')

### 3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [78]:
df_ords_2.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329       1             1                  2                  8   
1  2398795       1             2                  3                  7   
2   473747       1             3                  3                 12   
3  2254736       1             4                  4                  7   
4   431534       1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [79]:
## Changing the columns without overwriting the dataframe
df_ords_2.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = False)

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0        2539329       1             1                  2                  8   
1        2398795       1             2                  3                  7   
2         473747       1             3                  3                 12   
3        2254736       1             4                  4                  7   
4         431534       1             5                  4                 15   
...          ...     ...           ...                ...                ...   
3421078  2266710  206209            10                  5                 18   
3421079  1854736  206209            11                  4                 10   
3421080   626363  206209            12                  1                 12   
3421081  2977660  206209            13                  1                 12   
3421082   272231  206209            14                  6                 14   

         days_since_last_order  
0                          NaN  
1                         15.0  
2                         21.0  
3                         29.0  
4                         28.0  
...                        ...  
3421078                   29.0  
3421079                   30.0  
3421080                   18.0  
3421081                    7.0  
3421082                   30.0  

[3421083 rows x 6 columns]

### 4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [80]:
df_ords_2['order_hour_of_day'].value_counts()

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

#### Ans: The busiest hour for placing orders is 10am.

### 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [81]:
# Checking the data dictionary for what department 4 represent
print(data_dict.get('4'))

{'department': 'produce'}


#### Ans: The value 4 in the 'department_id' column represent the produce department.

### 6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [84]:
# Checking which department number represent 'breakfast'
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [85]:
# Create a df_breakfast dataframe with only product with 'department_id' of 14
df_breakfast = df_prods[df_prods['department_id'] == 14]

In [86]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

In [93]:
df_breakfast['department_id'].value_counts()

department_id
14    1116
Name: count, dtype: int64

### 7. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [87]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [106]:
# Create a df_dinner_parties dataframe with product with 'department_id' of 5(alcohol), 7(beverages), 12(meat seafood), 20(deli)
df_dinner_parties = df_prods[df_prods['department_id'].isin([5,7,12,20])]

In [107]:
df_dinner_parties.head(15)

product_id                                       product_name  aisle_id  \
2            3               Robust Golden Unsweetened Oolong Tea        94   
6            7                     Pure Coconut Water With Orange        98   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
16          17                                  Rendered Duck Fat        35   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   
43          44                        Sparkling Raspberry Seltzer       115   
48          49      Vegetarian Grain Meat Sausages Italian - 4 CT        14   
51          52                                Mirabelle Brut Rose       134   
52          53              Whole Leaf Pure Aloe With Lemon Juice        98   
61          62            Premium Deli Oven Roasted Turkey Breast        96   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1  
19              7     6.0  
22             12     8.2  
34             12    15.1  
38              7    12.5  
39             12    22.5  
43              7    12.2  
48             20    10.1  
51              5    14.4  
52              7     4.7  
61             20    14.6

In [108]:
df_dinner_parties['department_id'].value_counts()

department_id
7     4365
20    1322
5     1056
12     907
Name: count, dtype: int64

### 8. How many rows does the last dataframe you created have?

In [109]:
# Finding the number of columns and row in df_dinner_parties
df_dinner_parties.shape

(7650, 5)

#### Ans: 7650 rows.

### 9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [110]:
# Create a df_user_1 dataframe with user_id of 1
df_user_1 = df_ords_2[df_ords_2['user_id'] == '1']

In [113]:
df_user_1

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   
5   3367565       1             6                  2                  7   
6    550135       1             7                  1                  9   
7   3108588       1             8                  1                 14   
8   2295261       1             9                  1                 16   
9   2550362       1            10                  4                  8   
10  1187899       1            11                  4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

### 10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [114]:
df_user_1.describe()

order_number  order_day_of_week  order_hour_of_day  \
count     11.000000          11.000000          11.000000   
mean       6.000000           2.636364          10.090909   
std        3.316625           1.286291           3.477198   
min        1.000000           1.000000           7.000000   
25%        3.500000           1.500000           7.500000   
50%        6.000000           3.000000           8.000000   
75%        8.500000           4.000000          13.000000   
max       11.000000           4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

### 12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [117]:
#Exporting df_ords_2 dataframe
df_ords_2.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

#### 13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [118]:
#Exporting df_dep_t_new dataframe
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))